### ISSUE TITLE
关于ILIF在实现上的几个疑问

### ISSUE THREAD

**Issue by 2ephyrus** [BICLab/SpikeYOLO#70]
整数训练二进制推理是一个伟大的创新，从量化误差角度考虑，这无疑会给SNN训练带来性能提升。然而，我在神经形态数据集对ILIF进行实验取得了糟糕的结果。首先，我想询问您代码中的一处问题。
class mem_update 中的这段代码：
`       if i >= 1:
                mem = (mem_old - spike.detach()) * decay + x[i]
            else:
                mem = x[i]`
为什么spike要进行detach？这在神经元以往的工作中是罕见的，能否解释下这样做的目的？是为了便于训练得到更优的效果，还是从原理上应该进行detach？

我基于spikingjelly，在图像领域的SNN代码框架下实现了ILIF神经元，具体如下：
```
class ILIF(LIFNode_sj):

    def __init__(self, tau: float = 2., decay_input: bool = False, v_threshold: float = 1.,
                 v_reset: float = None, surrogate_function: Callable = Clip(),
                 detach_reset: bool = False, cupy_fp32_inference=False, channel=0, **kwargs):
        super().__init__(tau, decay_input, v_threshold, v_reset, surrogate_function, detach_reset, cupy_fp32_inference)
    def forward(self, x: torch.Tensor):
        # Leaky and Integrate
        self.v = self.v * 0.25 + x
        # Spike
        spike = self.surrogate_function(self.v)
        # reset
        self.v = self.v - spike
        return spike
```

```
class Clip(nn.Module):

    class quant4(torch.autograd.Function):

        @staticmethod
        def forward(ctx, input):
            ctx.save_for_backward(input)
            return torch.clamp(torch.round(input), min=0, max=4)

        @staticmethod
        def backward(ctx, grad_output):
            input, = ctx.saved_tensors
            grad_input = grad_output.clone()
            #             print("grad_input:",grad_input)
            grad_input[input < 0] = 0
            grad_input[input > 4] = 0
            return grad_input

    def forward(self, x):
        return self.quant4.apply(x)
```
我在CIFAR10-DVS数据集上进行了短期实验，结果表明，LIF+rectangle的效果显著优于I-LIF，这使我十分困惑。此外，I-LIF的位数越多，并没有得到理想中的性能提升，反而出现了退步。这令我十分困惑。我使用的框架为spiking_vgg11。
请问作者，这是否说明I-LIF不适用于神经形态数据？或者我使用的架构不合适？还是说我的神经元超参数设置（方程有误？）不合理？

作者的工作十分创新，并启发我进一步探索，因此我十分期待作者的回复，万分感谢！

**Comment by XinhaoLuo666** [comment]
感谢您的支持！关于您的第一个问题，spike的detach()并不是必须的。在我们的后续实验中，这似乎并不影响结果。而第二个问题，我认为是CIFAR10-DVS的任务相对简单。I-LIF在更为复杂的任务(尤其是RGB任务上，例如Imagenet)，或复杂的dvs任务(har-dvs，gen1)等具有更好的效果。而针对简单任务，普通的神经元足以将量化误差缩小到可接受的范围内。
